src/material-experimental/mdc-slider/global-change-and-input-listener.ts

```diff
@@ -27,44 +27,27 @@ export class GlobalChangeAndInputListener<K extends 'change'|'input'> {
   private _document: Document;
 
   /** Stores the subjects that emit the events that occur on the global document. */
-  private _subjects = new Map<K, Subject<Event>>();
-
-  /** Stores the event handlers that emit the events that occur on the global document. */
-  private _handlers = new Map<K, ((event: Event) => void)>();
+  private _observables = new Map<K, Observable<Event>>();
 
   constructor(@Inject(DOCUMENT) document: any, private _ngZone: NgZone) {
     this._document = document;
   }
 
-  /** Returns a function for handling the given type of event. */
-  private _createHandlerFn(type: K): ((event: Event) => void) {
-    return (event: Event) => {
-      this._subjects.get(type)!.next(event);
-    };
-  }
-
   /** Returns a subscription to global change or input events. */
   listen(type: K, callback: SpecificEventListener<K>): Subscription {
-    // This is the first subscription to these events.
-    if (!this._subjects.get(type)) {
-      const handlerFn = this._createHandlerFn(type).bind(this);
-      this._subjects.set(type, new Subject<Event>());
-      this._handlers.set(type, handlerFn);
-      this._ngZone.runOutsideAngular(() => {
-        this._document.addEventListener(type, handlerFn, true);
-      });
+    // If this is the first time we are listening to this event, create the observable for it.
+    if (!this._observables.has(type)) {
+      const observable = fromEvent(this._document, type, {capture: true}).pipe(
+        share(),
+        finalize(() => this._observables.delete(type)),
```

Comment:
I think instead of this we want:
```ts
private _destroyed = new Subject();

ngOnDestroy() {
	this._destroyed.next();
	this._destroyed.complete();
    this._observables().clear();
}

...pipe(
	takeUntil(this._destroyed),
	share()),
```

Reply:
Won't this cause a memory leak? What can trigger this `ngOnDestroy` except navigating away from the page?

Reply:
This is what we do in our other injectable services. The actual listeners are already getting removed when people unsubscribe from them because of how `fromEvent` works, but the `takeUntil` is just extra insurance that if someone forgets to unsubscribe we'll still complete the observable when `ngOnDestroy` fires. The map is going to have at most 2 keys, so I'm not worried about deleting them, but if you want you can leave the `finalize` in too, just put it before the `share` to make sure it only runs once